### PR TITLE
Migrate to nREPL 0.4

### PR DESCRIPTION
**Pull request by bbatsov**
An initial stab at updating the middleware for nREPL 0.4. 

That's going to be part of the 0.19 release (or at least that's the current plan). 

Any volunteers to test with using https://github.com/nrepl/lein-nrepl? (use `lein install` to build the artefact locally)

### CODE REVIEW

**Comment by arrdem** [comment]
LGTM, will try to make time to test it locally.

**Comment by bbatsov** [comment]
The changes seem trivial, but for some reason this doesn't work right. When I try evaluation with nREPL 0.4 alone everything is fine, when I add the middleware I can connect to the server but afterwards evaluation locks up. 

So far I couldn't find what exactly is going wrong, so I'd certainly appreciate some help in testing this.

**Comment by bbatsov** [comment]
```(-->
  id         "7"
  op         "out-subscribe"
  session    "0826418d-8cc2-44b9-9e20-15f917e6417c"
  time-stamp "2018-07-17 11:25:30.601503000"
)
(-->
  id           "8"
  op           "init-debugger"
  session      "0826418d-8cc2-44b9-9e20-15f917e6417c"
  time-stamp   "2018-07-17 11:25:30.615373000"
  print-length 10
  print-length 10
  print-level  10
  print-level  10
)
(<--
  id            "7"
  session       "0826418d-8cc2-44b9-9e20-15f917e6417c"
  time-stamp    "2018-07-17 11:25:30.689880000"
  out-subscribe "0826418d-8cc2-44b9-9e20-15f917e6417c"
  status        ("done")
)
(-->
  id           "9"
  op           "eval"
  session      "0826418d-8cc2-44b9-9e20-15f917e6417c"
  time-stamp   "2018-07-17 11:26:18.215710000"
  code         "10"
  column       7
  content-type "true"
  file         #("*cider-repl localhost(clj)*" 12 21 (ivy-index 0))
  line         45
  ns           "user"
)
```

That's what happens for me. Everything in the init sequence seems fine. My only guess is that init-debug might be causing some lockup but I haven't checked this yet.

Such a mystery...

**Comment by bbatsov** [comment]
Actually it seems that some things continue to be broken even with track-state disabled (e.g. eldoc). That's one hell of a mystery!

**Comment by bbatsov** [comment]
Likely last update from me for today, as this got me extremely pissed off - seemed like some problem with `info`, but I can't figure out what exactly's going on. Requests are just timing out, no errors are raised from the middleware and I have been unable to debug it, as the damn lein plugin I wrote is throwing today some bizarre class not found error, when I try to run it outside its project. What was supposed to be an easy tasks managed to frustrate me on an epic scale...

**Comment by dpsutton** [comment]
One lesson that i learned from the cemerick -> cider transition for piggieback:

Let's make a release of the new nrepl under the new namespaces that is _identical_ in all respects except for the namespaces. This makes sure that everyone can move to the new nrepl and the tooling doesn't need to case on one or the other. that was a pain.

**Comment by bbatsov** [comment]
No problem. Obviously dealing with this is not my dream in life, but it's the only way to fix the fundamental problems. Otherwise we can just pack up shop even now, because I'm done dealing with the limitations of the legacy nREPL and I'd rather just abandon the work on CIDER completely instead of wasting my time. 

Btw, if you look at the version in the branch you'll see it's 0.19, instead of 0.18 because I want to work on this in parallel with 0.18 and give people so time to adjust other related packages. 

I'm well aware there'll be some frustration after this release, but I didn't think the frustration would be not being able to even update the middlewares. :-)

**Comment by dpsutton** [comment]
```bash
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-apropos, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-classpath, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-complete, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-debug, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-enlighten, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-format, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-info, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-inspect, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-macroexpand, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-slurp, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-ns, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-out, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-content-type, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-slurp, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-pprint, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-pprint-fn, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-profile, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-refresh, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-resource, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-spec, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-stacktrace, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-test, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-trace, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-tracker, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-undef, see clojure.tools.middleware/set-descriptor!
[WARNING] No nREPL middleware descriptor in metadata of #'cider.nrepl/wrap-version, see clojure.tools.middleware/set-descriptor!
nREPL server started on port 38431 on host :: - nrepl://:::38431
Jul 18, 2018 12:15:49 AM clojure.tools.logging$eval429$fn__432 invoke
SEVERE: Unhandled REPL handler exception processing message {:op out-subscribe, :session 7c5592c7-43f3-4aff-9f23-04acbe3c8faf, :id 7}
java.lang.IllegalAccessError: newline? is not public, compiling:(clojure/tools/reader/reader_types.clj:1:1)
```

lots of info in the `*nrep-server fizzbuzz*` window for me.

**Comment by dpsutton** [comment]
and that last line is very strange:
`java.lang.IllegalAccessError: newline? is not public, compiling:(clojure/tools/reader/reader_types.clj:1:1)`

**Comment by bbatsov** [comment]
@dpsutton Are you start nREPL 0.2 or 0.4? You'd be seeing such warnings with 0.2, but not with 0.4.

**Comment by dpsutton** [comment]
```bash
[dan@fedora fizzbuzz]$ lein nrepl :middleware "['cider.nrepl/cider-middleware']"
(:middleware ['cider.nrepl/cider-middleware'])
Exception in thread "main" java.lang.ClassNotFoundException: leiningen.nrepl, compiling:(/tmp/form-init7845473687617269713.clj:1:73)

```

not sure what's up with that. i'll have to look in the morning. its quite late now. I'm gonna dig into this with you @bbatsov . This sucks its being so strange.

**Comment by bbatsov** [comment]
> not sure what's up with that.

Seems like a bug in the plugin I've created, although I can't figure this one out as well. It's just a super standard one file lein plugin.

File: src/cider/nrepl/middleware/stacktrace.clj [hunk 1]
```diff
@@ -110,7 +110,7 @@
 
 (defn flag-tooling
   "Walk the call stack from top to bottom, flagging frames below the first call
-  to `clojure.lang.Compiler` or `clojure.tools.nrepl.*` as `:tooling` to
+  to `clojure.lang.Compiler` or `nrepl.*` as `:tooling` to
   distinguish compilation and nREPL middleware frames from user code."
   [frames]
   (let [tool-regex #"^clojure\.lang\.Compiler|^clojure\.tools\.nrepl|^cider\."
```
Comment (tatut) [inline src/cider/nrepl/middleware/stacktrace.clj]: Shouldn't the `clojure\.tools\.nrepl`  in the regex also be changed to just `nrepl`?

Comment (bbatsov) [reply src/cider/nrepl/middleware/stacktrace.clj]: Yeah, it should be changed.

**Comment by reedho** [comment]
I got same error as above testing with https://github.com/nrepl/lein-nrepl

```
 /p/t/randos $ lein with-profile production nrepl
nil
Exception in thread "main" java.lang.ClassNotFoundException: leiningen.nrepl, compiling:(/private/var/folders/vt/k15mc3ln5vx1n7_9x9mqh7q40000gn/T/form-init14967222619861060167.clj:1:126)
```

Exploring other option, using clj deps (without lein) seems there is no problem at all, can connect and repl-ing normally, midleware work as expected. The gist i put here: https://gist.github.com/reedho/3d3aeb805f20fc6f3aec03881c74cab4

My best guess is that `lein` always injecting old `org.clojure/tools.nrepl` for non `production` profile.

**Comment by dpsutton** [comment]
I managed to get the contents of that temp file that is throwing the error:

```
(.deleteOnExit (java.io.File. "/tmp/form-init2271818862159530970.clj")) (do (set! *warn-on-reflection* nil) nil (leiningen.nrepl/start-nrepl {:middleware [(quote cider.nrepl/cider-middleware')]}))
```

**Comment by dpsutton** [comment]
to contrast, other tmp files have

```clojure
(try (clojure.core/require (quote cider.nrepl)) (catch java.lang.Throwable t__11809__auto__ (clojure.core/println "Error loading" (clojure.core/str (quote cider.nrepl) ":") (clojure.core/or (.getMessage t__11809__auto__) (clojure.core/type t__11809__auto__)))))
```
 in them. not sure why these have a created require message in them but that would explain the "what the heck is leiningen.repl" message we are seeing

**Comment by dpsutton** [comment]
> Inside the eval-in-project call the project's own classpath will be active and Leiningen's own internals and plugins will not be available.

https://github.com/technomancy/leiningen/blob/master/doc/PLUGINS.md#evaluating-in-project-context

this seems to be an abuse of eval in project apparently.

**Comment by bbatsov** [comment]
@dpsutton Yeah, I also figured this out in the morning (and not just me https://github.com/nrepl/lein-nrepl/pull/1). I really wished this produced a more meaningful error, though.  Now I'm wondering what's the best course of action. Probably it's best to add this dep if it's not in the project deps, but I'm not a Lein expert and I don't know what's the best strategy there.

**Comment by bbatsov** [comment]
> My best guess is that lein always injecting old org.clojure/tools.nrepl for non production profile.

@reedho Wow! You're probably right! Lein has a hard dep on the old nREPL, but I didn't think it could mess something up because of the different namespaces. Likely, however, that's not completely true.

**Comment by SevereOverfl0w** [comment]
> I didn't think it could mess something up because of the different namespaces. Likely, however, that's not completely true.

I'm not sure what else could collide here? I suppose this can easily be verified via the `clj` tools.

**Comment by SevereOverfl0w** [comment]
```
❯ clj -Sdeps '{:deps {nrepl {:mvn/version "0.4.1"} org.clojure/tools.nrepl {:mvn/version "0.2.13"}}}' -m nrepl.cmdline --interactive
network-repl
Clojure 1.9.0
user=> (+ 1 1)
2
```

Nope, you can have new & old nrepl alongside one another.

**Comment by bbatsov** [comment]
> I'm not sure what else could collide here? I suppose this can easily be verified via the clj tools.

@SevereOverfl0w Yeah, that's what I plan to do when I find some time. But I did notice yesterday I can't jack in the nREPL project, so I guess something goes sideways when the two jars nREPLs are present.

**Comment by bbatsov** [comment]
@SevereOverfl0w For me the problems started only after adding the `cider-nrepl` middleware (and not everything was broken, just some things). I guess you should also load some middleware and see if var-info is working.

**Comment by bbatsov** [comment]
`clj -Sdeps '{:deps {cider/cider-nrepl {:mvn/version "0.19.0-SNAPSHOT"} }}' -e '(require (quote cider-nrepl.main)) (cider-nrepl.main/init ["cider.nrepl/cider-middleware"])'` works like a charm. All the middleware are fine. 

Guess @reedho is right and something's getting messed up because of lein's nREPL dep. At least we're making progress. Hopefully with Boot there will be no such problems (I'll leave it to @arichiardi to test this). Now we have to figure out exactly what's going wrong with `lein nrepl`.

**Comment by SevereOverfl0w** [comment]
I don't think it is because of the nrepl dep:

```
❯ clj -Sdeps '{:deps {nrepl {:mvn/version "0.4.1"} org.clojure/tools.nrepl {:mvn/version "0.2.13"} cider/cider-nrepl {:mvn/version "0.17.0"}}}' -e '(require (quote cider-nrepl.main)) (cider-nrepl.main/init ["cider.nrepl/cider-middleware"])'
nREPL server started on port 35557 on host 0:0:0:0:0:0:0:0 - nrepl://0:0:0:0:0:0:0:0:35557
```

**Comment by bbatsov** [comment]
Well, for some reason it works fine like this and it doesn't work with `lein nrepl` - what else can I think? :-)

**Comment by bbatsov** [comment]
Yep, seems we're all good. I haven't tested only the boot support, but probably everything's fine there as well.

I'll write some notes about using this with CIDER and I'll leave this open for now, until 0.18 is released. @vspinu asked us to postpone the release a bit. In the mean time I'll update piggieback (likely to support both nREPLs) and ping @ttheller and @bhauman to add some support for nREPL 0.4 as well. The changes they'll have to make to support both nREPLs should be small.

**Comment by bbatsov** [comment]
Big thanks to everyone who helped with this!

In case someone's curious what the problem turned out to be - the short story is that I wrote a very buggy lein plugin. :-) I really need to read up on Leiningen one of those days...

**Comment by arichiardi** [comment]
Confirm it works with `nrepl-server` and https://github.com/clojure-emacs/cider/pull/2379 applied.

**Comment by bbatsov** [comment]
The actual work went into 0.18 (https://github.com/clojure-emacs/cider-nrepl/commit/15d87747f790d300b0f13eef2e5bdc87b9dfd4ec)

**Pull request merged**